test/integration/daemonset/daemonset_test.go

```diff
@@ -992,3 +993,78 @@ func TestUnschedulableNodeDaemonDoesLaunchPod(t *testing.T) {
 		validateDaemonSetStatus(dsClient, ds.Name, 2, t)
 	})
 }
+
+func TestUpdateStatusDespitePodCreationFailure(t *testing.T) {
+	forEachStrategy(t, func(t *testing.T, strategy *apps.DaemonSetUpdateStrategy) {
+		limitedPodNumber := 2
+		// use framework.StartTestServer to inject admission control
+		// e.g. to emulate the resource quota behavior
+		c, config, closeFn := framework.StartTestServer(t, framework.TestServerSetup{
+			ModifyServerConfig: func(config *controlplane.Config) {
+				config.GenericConfig.AdmissionControl = &fakePodFailAdmission{
+					limitedPodNumber: limitedPodNumber,
+				}
+			},
+		})
+		defer closeFn()
+
+		resyncPeriod := 12 * time.Hour
+		informers := informers.NewSharedInformerFactory(clientset.NewForConfigOrDie(restclient.AddUserAgent(config, "daemonset-informers")), resyncPeriod)
+		dc, err := daemon.NewDaemonSetsController(
+			informers.Apps().V1().DaemonSets(),
+			informers.Apps().V1().ControllerRevisions(),
+			informers.Core().V1().Pods(),
+			informers.Core().V1().Nodes(),
+			clientset.NewForConfigOrDie(restclient.AddUserAgent(config, "daemonset-controller")),
+			flowcontrol.NewBackOff(5*time.Second, 15*time.Minute),
+		)
+		if err != nil {
+			t.Fatalf("error creating DaemonSets controller: %v", err)
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		eventBroadcaster := events.NewBroadcaster(&events.EventSinkImpl{
```

Comment:
@gjkim42  looking at this again, I think it is a bit unfortunate that we are duplicating code from `setup` here.  I know we need to modify the server and I am not seeing such option in the `setup`'s apiserver.  Can we pass a startServer function to the setup that could be used instead of the original one if specified? I suppose it could be also solved declaratively, but the functional approach seems more versatile. 

This way, we can share all of the setup intentions and this test can react to the future changes of setup.

Reply:
@atiratree 

> Can we pass a startServer function to the setup that could be used instead of the original one if specified?

To do this, replaced `cmd/kube-apiserver/app/testing.StartTestServer` with `test/integration/framework.StartTestServer` and added `setupWithServerSetup` to be able to configure the admission control of the apiserver.

Reply:
@gjkim42 thank you - looks much better now